torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -1744,3 +1745,31 @@ def test_use_rref_after_shutdown(self):
         ):
             import torch.distributed.rpc.internal as internal
             internal.serialize(rref)
+
+@unittest.skipIf(
+    not torch._six.PY3,
+    "Pytorch distributed autograd package " "does not support python2",
+)
+class FaultyAgentRpcTest(RpcAgentTestFixture):
+
+    # no faulty_messages defined so this fails all retryable messages - see
+    # test/distributed/rpc/faulty_agent/test_faulty_agent_rpc_trap.py for the
+    # list of retryable messages.
+    @dist_init
+    def test_check_failed_messages(self):
+        if self.rank == 0:
+            dst_worker_b = "worker{}".format((self.rank + 1) % self.world_size)
+            dst_worker_c = "worker{}".format((self.rank + 2) % self.world_size)
+
+            # Worker0 sends RPC to Worker1 and creates an RRef there
+            rref = rpc.remote(dst_worker_b, torch.add, args=(torch.ones(2, 2), torch.ones(2, 2)))
+            # Worker0 sends an RPC to Worker2 with the RRef as an arg
+            rpc.remote(dst_worker_c, add_rref_to_value, args=(rref, torch.ones(2, 2)))
+            # remove reference to Worker0's UserRRef
+            rref = None
+        # explicitly delete all User RRefs
+        _delete_all_user_rrefs()
+        # wait for all the deletion messages to complete before the test
+        # calls shutdown
+        wait_until_pending_forks_flushed()
```

Comment:
does this mean applications also need to do this if using retry? Any reason we don't do this in our `rpc.shutdown`?

BTW, since #31893, `_destroy_rref_context()` in `shutdown()` will wait for all `UserRRef`s to be confirmed by owner and accepted by child. Does that change have any implication here?

Reply:
Thanks for the pointer @mrshenli! After rebasing, looks like `wait_until_pending_forks_flushed` and `wait_until_pending_forks_flushed` are no longer needed. This was only because we're purposely failing the `RREF_USER_DELETE` messages the first 3 times, and we will see an RRef leak if we don't ensure those messages are processed before/during shutdown, but seems like this is taken care of now.